File: interview_questions/ch2-Linked-lists.py


# class Node:
#     def __init__(self, data=None):
#         self.data = data
#         self.next = None

# class LinkedList:
#     def __init__(self):
#         self.head = Node()

#     def append(self, data):
#         new_node = Node(data)
#         current = self.head
#         while current.next != None:
#             current = current.next
#         current.next = new_node

#     def display(self):
#         elems = []
#         current_node = self.head
#         while current_node.next != None:
#             current_node = current_node.next
#             elems.append(current_node.data)
#         print(elems)
#         return elems

#     def removeDuplicate(self):
#         hashTable = {}
#         current_node = self.head
#         while current_node.next != None:
#             prev_node = current_node
#             current_node = current_node.next
#             if current_node.data in hashTable:
#                 prev_node.next = current_node.next 
#             else:
#                 hashTable[current_node.data] = True
#         print(hashTable)



class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None

    
class LinkedList:
    def __init__(self):
        self.head = Node()


    def addToList(self, data):
        new_node = Node(data)
        current_node = self.head
        while current_node.next != None:
            current_node = current_node.next
        current_node.next = new_node

    def display(self):
        elems = []
        current_node = self.head
        while current_node.next != None:
            current_node = current_node.next
            elems.append(current_node.data)
        print(elems)
        return elems

    def length(self):
        length = 0;
        current_node = self.head
        while current_node.next != None:
            length += 1
            current_node = current_node.next
        # print(length)
        return length

    def erase(self, index):
        current_index = 0;
        current_node = self.head
        while current_node.next != None:
            prev_node = current_node
            current_node = current_node.next
            if current_index == index:
                prev_node.next = current_node.next
            current_index += 1

    def removeDups(self):
        current_node = self.head
        hashTable = {}
        while current_node.next != None:
            prev_node = current_node
            current_node = current_node.next
            if current_node.data in hashTable:
                prev_node.next = current_node.next
            else:
                hashTable[current_node.data] = True;
        return self

    def get(self, index):
        current_node = self.head
        current_index = 0
        while current_node.next != None:
            current_node = current_node.next
            if current_index == index:
                print(current_node.data)
            current_index += 1
        
    def divideList(self, Linkedlist):
        if self.length() == 0:
            print("ERROR: List is empty")
            return None

        current_node = self.head
        node_runner = self.head

        while node_runner.next != None:
            node_runner = node_runner.next

            if node_runner.next == None:
                break

            node_runner = node_runner.next
            current_node = current_node.next

        toReturn = current_node.next
        current_node.next = None
        return toReturn
        



test = LinkedList();
test.addToList(4);
test.addToList(1);
test.addToList(2);
test.addToList(2);
test.addToList(3);
test.addToList(4);
test.addToList(2);
test.display();
test.divideList(test)
test.display();




































# my_list = LinkedList()

# my_list.append(2)
# my_list.append(16)
# my_list.append(4)
# my_list.append(2)
# my_list.append(6)
# my_list.append(2)
# my_list.append(27)
# my_list.append(29)
# my_list.append(16)
# my_list.append(22)
# my_list.append(2)
# my_list.append(77)
# my_list.append(2)
# my_list.append(27)
# my_list.append(29)
# my_list.append(16)
# my_list.append(22)

# my_list.display()

# my_list.removeDuplicate()
# my_list.removeDuplicate()
# my_list.removeDuplicate()
# my_list.display()

File: interview_questions/ch2/MISC/ch2-Linked-lists.js
class Node {
    constructor(data){
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = new Node();
    }

    addToHead(value) {
        const new_node = new Node(value);
        var current_node = this.head
        while(current_node.next !== null){
            current_node = current_node.next;
        }
        current_node.next = new_node;
    }

    length(){
        let current_node = this.head;
        var length = 0;
        while (current_node.next !== null){
            length += 1;
            current_node = current_node.next;
        }
        console.log(length)
        return length
    }

    display(){
        var list = []
        var current_node = this.head
        while (current_node.next !== null){
            current_node = current_node.next;
            list.push(current_node.data);
        }
        console.log(list)
    }

    get(index){
        if(index >= this.length()){
            let error = "The index is beyond the linked list count"
            console.log(error)
            return error;
        }
        let current_index = 0;
        let current_node = this.head;
        while(true){
            current_node = current_node.next
            if(current_index == index){
                console.log(current_node.data)
                return current_node.data
            }
            current_index += 1
        }
    }

    erase(index){
        if(index >= this.length()){
            let error = "The index is beyond the linked list count"
            console.log(error)
            return error;
        }
        let current_index = 0;
        let current_node = this.head;

        while(true){
            let last_node = current_node;
            current_node = current_node.next
            if(current_index == index){
                last_node.next = current_node.next
                return false
            }
            current_index += 1
        }
    }

    eraseDups(){
        let current_node = this.head;
        let hash = {}
        while(current_node.next !== null){
            let prev_node = current_node
            current_node = current_node.next
            if(current_node.data in hash){
                // test
                // prev_node 
                // end test
                // console.log('Number why it tripped the if >>>>> ',current_node.data);
                prev_node.next = current_node.next
            } else {
                hash[current_node.data] = true
                // prev_node.next = current_node.next
            }
        }
        // console.log(hash)
    }
};

var test = new LinkedList()
test.addToHead(1)
test.addToHead(2)
test.addToHead(3)
test.addToHead(3)
test.addToHead(4)
test.addToHead(5)
test.addToHead(3)
test.addToHead(33)
test.addToHead(3)
test.addToHead(34)
test.addToHead(3)
test.addToHead(3)

test.display()
test.eraseDups()
test.display()
